### PR TITLE
Onboard Camel K Runtime

### PR DESCRIPTION
**Pull request by claudio4j**
- This is a JVM only extension

The intention is to collect early feedback.

https://github.com/apache/camel-quarkus/issues/4283

There are some parts that requires more investigation, like the use an existing camel-quarkus Runtime interface, instead of the one in the k-core/api package
Also, there are some tests disabled due to failures in the CamelContextCustomizer tests, which I am investigating.
There is the ApplicationConfigSourceProvider which I will see if it is really necessary to depend on microprofile config.

### CODE REVIEW

**Comment by claudio4j** [comment]
I am looking at the failed validations.

**Review by squakez — COMMENTED** [review]
The concept of adding the camel-k-runtime as a camel quarkus module is good. Right now, the runtime release cycle is completely coupled to Camel Quarkus, so, it will simplify the release.

I can see it misses a few components such as camel-k-cron or camel-k-resume-kafka (not sure if this latest one is fully functional though, cc @orpiske). It also misses the support plugins (such as the catalog generator) or the platform bom which are used directly by Camel K to perform certain actions.

**Comment by orpiske** [comment]
> The concept of adding the camel-k-runtime as a camel quarkus module is good. Right now, the runtime release cycle is completely coupled to Camel Quarkus, so, it will simplify the release.
> 
> I can see it misses a few components such as camel-k-cron or camel-k-resume-kafka (not sure if this latest one is fully functional though, cc @orpiske). It also misses the support plugins (such as the catalog generator) or the platform bom which are used directly by Camel K to perform certain actions.

I haven't had a chance to try with camel k 2.x yet. AFAIK, we did not have any change on core that could directly affect this, so it should work (as in general - not sure about Camel K side of changes).

**Comment by claudio4j** [comment]
@squakez it is intentional to bring in only the camel-k-core and camel-k-runtime modules of the camel-k-runtime project to collect early feedback, about the project structure, naming, api usage, etc.
In the meantime I am working to bring the remaining modules.

**Comment by jamesnetherton** [comment]
For naming, I'm not sure I like prefixing stuff with `k`. I'd prefer we use `camel-k`, then there's no ambiguity. Do we want to name artifacts like `camel-quarkus-k-core-api` or just simplify to `camel-k-core-api`? What do others think?

Then there's the question of how we strucure things. We could either continue with what's proposed here with something like:

```
|- extensions-jvm
|----- camel-k
|--------- core
|--------- runtime
|--------- foo
|--------- bar
```

Or have dedicated top level modules for the camel-k specific stuff. E.g:

```
| - extensions-jvm-camel-k
|----- core
|----- runtime
```

For any native supported bits:

```
| - extensions-camel-k
|----- extension-a
|----- extension-b
```

And a similar split for the integration tests `integration-tests-jvm-camel-k` & `integration-tests-camel-k`. 

I am generally terrible at naming / deciding these things, so it'd be good for others to comment.

**Comment by zbendhiba** [comment]
> For naming, I'm not sure I like prefixing stuff with `k`. I'd prefer we use `camel-k`, then there's no ambiguity. Do we want to name artifacts like `camel-quarkus-k-core-api` or just simplify to `camel-k-core-api`? What do others think?
> 

I have a preference for `camel-quarkus-k-core-api`, unless there's a technical reason from Camel-k project.

> Then there's the question of how we strucure things. We could either continue with what's proposed here with something like:
> 
> ```
> |- extensions-jvm
> |----- camel-k
> |--------- core
> |--------- runtime
> |--------- foo
> |--------- bar
> ```
> 
> Or have dedicated top level modules for the camel-k specific stuff. E.g:
> 
> ```
> | - extensions-jvm-camel-k
> |----- core
> |----- runtime
> ```

This could be a first way of dealing with this.

> 
> For any native supported bits:
> 
> ```
> | - extensions-camel-k
> |----- extension-a
> |----- extension-b
> ```
> 
> And a similar split for the integration tests `integration-tests-jvm-camel-k` & `integration-tests-camel-k`.
> 
> I am generally terrible at naming / deciding these things, so it'd be good for others to comment.

**Comment by lburgazzoli** [comment]
> There is the ApplicationConfigSourceProvider which I will see if it is really necessary to depend on microprofile config.

This is likely a good candidate for a first integration with Camel and/or Camel Quarkus as the goal for the `ApplicationConfigSourceProvider` aws to load properties in a particular order so that we can control the priority of the properties and we can also scan all the mounted secrets/configmaps.

Camel has already some priority concepts backed in the `PropertiesComponent`:
- default properties
- override properties
- all the others

So we can probably model what `ApplicationConfigSourceProvider` does with plain Camel.
In addition to that we can add the ability to scan directories Camel Quarkus in a form of a `ConfigSourceProvider`

File: extensions-jvm/k/k-core/api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            <artifactId>camel-endpointdsl</artifactId>
+        </dependency>
+
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
```
[the remark this answers is not in the document]
Comment (claudio4j) [reply extensions-jvm/k/k-core/api/pom.xml]: Done.

**Comment by claudio4j** [comment]
- Rebased main 
- Bring in the catalog from camel-k-runtime tooling. The maven plugin generates a specific camel-catalog.yaml to be consumed by camel k. it is the supported set of camel components.

File: extensions-jvm/k/bom/pom.xml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    <packaging>pom</packaging>
+
+    <properties>
+        <quarkus-platform-group>io.quarkus.platform</quarkus-platform-group>
+    </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>${quarkus-platform-group}</groupId>
```
Comment (jamesnetherton) [inline extensions-jvm/k/bom/pom.xml:41]:
```suggestion
                <groupId>org.apache.camel.quarkus</groupId>
```

File: extensions-jvm/k/bom/pom.xml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                <artifactId>quarkus-camel-bom</artifactId>
```
Comment (jamesnetherton) [inline extensions-jvm/k/bom/pom.xml:42]:
```suggestion
                <artifactId>camel-quarkus-bom</artifactId>
```

File: extensions-jvm/k/bom/pom.xml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                <version>${quarkus.version}</version>
```
Comment (jamesnetherton) [inline extensions-jvm/k/bom/pom.xml:43]:
```suggestion
                <version>${project.version}</version>
```

File: extensions-jvm/k/bom/pom.xml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+            <dependency>
+                <groupId>${quarkus-platform-group}</groupId>
```
Comment (jamesnetherton) [inline extensions-jvm/k/bom/pom.xml:48]:
```suggestion
                <groupId>io.quarkus</groupId>
```

**Comment by claudio4j** [comment]
I will rebase and push more changes later today.

**Comment by claudio4j** [comment]
@ppalaga can you have a look at this preliminary work ?

**Comment by ppalaga** [comment]
image looks scarry, but I will have a look later today

**Comment by ppalaga** [comment]
We had a chat with @claudio4j and we agreed to work towards the following (@jamesnetherton and others please speak up if you do not like something):

1. The BOM currently in `extensions-jvm/k/bom/pom.xml` will be eliminated by moving the entries to `poms/bom/pom.xml`. I believe it is better both for CQ maintainers and for Camel K maintainers to have just a single BOM. The newly added entries won't be many and they will hopefully not confuse Camel Quarkus users too much.
2. Some modules can perhaps be merged into others, e.g. there is a chance to have just a single camel-k specific extension. 
3. We need to figure out whether the new extension(s) needs to work in native mode (any opinion on that @lburgazzoli ?) and possibly move it from `extensions-jvm` to `extensions`. Same for the tests. Some tests will possibly have to be split into native/jvm only parts, because some DSL are only supported in JVM.
4. The intermediate `/k/` directory should be eliminated.

Besides that, I am fine with the current `camel-quarkus-k` name prefix, as well as I'd be fine with `camel-quarkus-camel-k` or `camel-quarkus-camelk`.

**Comment by claudio4j** [comment]
Thanks @ppalaga I am working to materialize these changes in this PR.

**Comment by claudio4j** [comment]
Rebased and removed the camel-k bom and camel-catalog modules.

**Review by ppalaga — COMMENTED** [review]
1. Thanks for adjusting the directory structure and BOM
2. `snakeyaml-engine` is excluded on several locations in the poms. I wonder why is it required? I see it coming via `kubernetes-client` and `kubernetes-client-api`. We do use `kubernetes-client` elsewhere, but we do not exclude the `snakeyaml-engine` there. I guess we should do the same thing everywhere unless there is a reason.
3. We need an authoritative opinion whether the new extension should support native mode. 

More inline

File: extensions-jvm/camel-k/deployment/pom.xml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        <dependency>
+            <groupId>org.apache.camel.quarkus</groupId>
+            <artifactId>camel-quarkus-bean-deployment</artifactId>
+        </dependency>
+        <!--
+        TODO: verify if these are needed
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-logging-json-deployment</artifactId>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (claudio4j) [reply extensions-jvm/camel-k/deployment/pom.xml]: yes, this dep is needed.

File: integration-tests-jvm/camel-k-loader/pom.xml [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                                        <quarkus.http.test-port>${test.http.port.native}</quarkus.http.test-port>
+                                        <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
+                                    </systemProperties>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
```
[the remark this answers is not in the document]
Comment (claudio4j) [reply integration-tests-jvm/camel-k-loader/pom.xml]: moved to integration-tests

File: integration-tests-jvm/camel-k-loader/pom.xml [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                    <version>${project.version}</version>
+                    <type>pom</type>
+                    <scope>test</scope>
+                    <exclusions>
+                        <exclusion>
+                            <groupId>*</groupId>
+                            <artifactId>*</artifactId>
+                        </exclusion>
+                    </exclusions>
+                </dependency>
```
[the remark this answers is not in the document]
Comment (claudio4j) [reply integration-tests-jvm/camel-k-loader/pom.xml]: done

**Comment by claudio4j** [comment]
The `snakeyaml-engine` exclusion was removed, it was added when the first I added the tests to camel-quarkus, but over the course of updates, the exclusion is not necessary anymore.

**Comment by ppalaga** [comment]
> I am still looking whether we could merge the ITs into fewer Maven modules to save some time on the CI.

Now I see that `camel-k-loader` does not have any assertions for native mode (there is no `*IT` class). Also based on what the tests are doing, I'd vote for transforming it into a lightweight deployment test -  a PR will follow.

`camel-k-runtime-xml` and `camel-k-runtime-yaml` could perhaps be merged together. Although we normally strive to test extensions with [minimal set of ancillary dependencies](https://camel.apache.org/camel-quarkus/2.16.x/contributor-guide/extension-testing.html#_minimal_set_of_dependencies) (which would be a reason to keep the two test modules separate), but in this specific case, we can say that (1) XML and YAML DSLs are tested in isolation elsewhere and (2) the Camel K extension is rather unlikely to be "fixed" by any of the two. Let me try to merge them together.

**Comment by claudio4j** [comment]
In camel-k-runtime repository, these tests are IT enabled. 
https://github.com/apache/camel-k-runtime/tree/main/itests/camel-k-itests-loader-xml/src/test/java/org/apache/camel/k/loader/xml/quarkus

It was probably some oversight of me not bringing them, which I can do now.

All of these tests are contained in their own project in camel-k-runtime, so when I brought them to camel-quarkus, I wanted to have them in the same project to save IT test time.

**Comment by ppalaga** [comment]
@claudio4j in https://github.com/apache/camel-quarkus/pull/5219 I have rebased on top of recent main and I have put a few enhancements on top of the current PR. Please review there and possibly amend to your PR.

**Comment by claudio4j** [comment]
Let's close this PR and work on #5219

**Pull request closed without merging**